ic-asset/src/asset_config.rs

```diff
@@ -0,0 +1,627 @@
+        if let Ok(glob) = Glob::new(&format!("{}/{}", dir.to_str().unwrap(), &glob)) {
+            Ok(glob.compile_matcher())
+        } else {
+            Err(serde::de::Error::custom(
+                "the value in `match` field is not a valid glob pattern",
+            ))
+        }
+    } else {
+        Err(serde::de::Error::custom(
+            "the value in `match` field is not a string",
+        ))
+    }
+}
+
+impl AssetConfigRule {
+    fn applies(&self, cannonical_path: &Path) -> bool {
+        self.r#match.is_match(cannonical_path)
+    }
+}
+
+impl AssetConfig {
+    fn merge(mut self, other: Self) -> Self {
+        if let Some(c) = other.cache {
+            self.cache = Some(c);
+        };
+        match (self.headers.as_mut(), other.headers) {
+            (Some(sh), Some(oh)) => sh.extend(oh),
+            (_, oh) => self.headers = oh,
+        };
+        self
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct AssetSourceDirectoryConfiguration {
+    config_map: Map,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize)]
+pub(crate) struct AssetConfig {
+    pub(crate) cache: Option<CacheConfig>,
+    pub(crate) headers: Option<HeadersConfig>,
+}
+
+impl Default for AssetConfig {
+    fn default() -> Self {
+        Self {
+            headers: Some(HashMap::new()),
+            cache: Some(CacheConfig::default()),
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct AssetConfigTreeNode {
+    pub parent: Option<Arc<AssetConfigTreeNode>>,
+    pub rules: Vec<AssetConfigRule>,
+}
+
+impl AssetSourceDirectoryConfiguration {
+    pub(crate) fn load(root_dir: &Path) -> anyhow::Result<Self> {
+        let mut config_map = HashMap::new();
+        AssetConfigTreeNode::load(None, &root_dir, &mut config_map)?;
+        Ok(Self { config_map })
+    }
+
+    pub(crate) fn get_asset_config(&self, canonical_path: &Path) -> AssetConfig {
+        self.config_map
+            .get(canonical_path.parent().unwrap())
+            .unwrap()
+            .get_config(canonical_path)
+    }
+}
+
+impl AssetConfigTreeNode {
+    fn load(
+        parent: Option<Arc<AssetConfigTreeNode>>,
+        dir: &Path,
+        configs: &mut HashMap<PathBuf, Arc<AssetConfigTreeNode>>,
+    ) -> anyhow::Result<()> {
+        let mut rules = vec![];
+        if let Ok(file) = File::open(dir.join(ASSETS_CONFIG_FILENAME)) {
+            let reader = BufReader::new(file);
+            CURRENTLY_PROCESSED_ASSETS_DIRECTORY.with(|book| {
+                let x = book.borrow_mut();
+                let mut path = RefMut::map(x, |p| p);
+                *path = dir.to_path_buf();
+            });
+
+            match serde_json::from_reader(reader) {
+                Ok(mut v) => rules.append(&mut v),
+                Err(e) => bail!("{:?}: {:?}", dir, e),
+            }
+        }
+
+        let config_tree = Self { parent, rules };
+        let parent_ref = Arc::new(config_tree);
```

Comment:
I think we need to have `AssetConfigTreeNodes` for all directories, no matter if they are empty, otherwise, the recursive function won't be able to recurse into top-_top_ parent directory

Reply:
As discussed, added `.is_some()` check above

Reply:
done in https://github.com/dfinity/agent-rs/pull/352/commits/ae2933a2e73f2aa903ccb10f27853efe77215bf5